File: android/app/src/main/java/com/locallink/android/LocalLinkCrypto.kt
package com.locallink.android

import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.modes.ChaCha20Poly1305
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import java.security.SecureRandom

class SessionCrypto(
  localPrivateKey: ByteArray,
  remotePublicKeyBase64: String
) {
  private val key: ByteArray

  init {
    val privateKey = X25519PrivateKeyParameters(localPrivateKey, 0)
    val remotePublicKey = X25519PublicKeyParameters(LocalLinkHashes.unbase64(remotePublicKeyBase64), 0)
    val agreement = X25519Agreement()
    agreement.init(privateKey)
    val shared = ByteArray(agreement.agreementSize)
    agreement.calculateAgreement(remotePublicKey, shared, 0)

    key = ByteArray(32)
    HKDFBytesGenerator(SHA256Digest()).apply {
      init(
        HKDFParameters(
          shared,
          "LocalLink-v1".toByteArray(Charsets.UTF_8),
          "frames".toByteArray(Charsets.UTF_8)
        )
      )
      generateBytes(key, 0, key.size)
    }
  }

  fun seal(payload: ByteArray): ByteArray {
    val nonce = ByteArray(12)
    SecureRandom().nextBytes(nonce)
    val cipher = ChaCha20Poly1305()
    cipher.init(true, AEADParameters(KeyParameter(key), 128, nonce))
    val encrypted = ByteArray(cipher.getOutputSize(payload.size))
    val len = cipher.processBytes(payload, 0, payload.size, encrypted, 0)
    cipher.doFinal(encrypted, len)
    return nonce + encrypted
  }

  fun open(sealed: ByteArray): ByteArray {
    require(sealed.size >= 12)
    val nonce = sealed.copyOfRange(0, 12)
    val encrypted = sealed.copyOfRange(12, sealed.size)
    val cipher = ChaCha20Poly1305()
    cipher.init(false, AEADParameters(KeyParameter(key), 128, nonce))
    val output = ByteArray(cipher.getOutputSize(encrypted.size))
    val len = cipher.processBytes(encrypted, 0, encrypted.size, output, 0)
    val finalLen = cipher.doFinal(output, len)
    return output.copyOf(len + finalLen)
  }

  companion object {
    fun makeIdentity(displayName: String): LocalIdentity {
      val privateKey = X25519PrivateKeyParameters(SecureRandom())
      val privateBytes = ByteArray(32)
      privateKey.encode(privateBytes, 0)
      val publicBytes = ByteArray(32)
      privateKey.generatePublicKey().encode(publicBytes, 0)
      val identity = DeviceIdentity(
        deviceID = java.util.UUID.randomUUID().toString(),
        displayName = displayName,
        publicKey = LocalLinkHashes.base64(publicBytes)
      )
      return LocalIdentity(identity, privateBytes)
    }
  }
}

File: windows/LocalLink.Windows/LocalLinkCrypto.cs
using System.Security.Cryptography;
using Org.BouncyCastle.Crypto.Agreement;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Security;
using BouncyCastleChaCha20Poly1305 = Org.BouncyCastle.Crypto.Modes.ChaCha20Poly1305;

namespace LocalLink.Windows;

public sealed class SessionCrypto
{
    private readonly byte[] key;

    public SessionCrypto(byte[] localPrivateKey, string remotePublicKeyBase64)
    {
        var privateKey = new X25519PrivateKeyParameters(localPrivateKey, 0);
        var publicKey = new X25519PublicKeyParameters(Convert.FromBase64String(remotePublicKeyBase64), 0);
        var agreement = new X25519Agreement();
        agreement.Init(privateKey);

        var shared = new byte[agreement.AgreementSize];
        agreement.CalculateAgreement(publicKey, shared, 0);

        key = new byte[32];
        var hkdf = new HkdfBytesGenerator(new Sha256Digest());
        hkdf.Init(new HkdfParameters(
            shared,
            "LocalLink-v1"u8.ToArray(),
            "frames"u8.ToArray()));
        hkdf.GenerateBytes(key, 0, key.Length);
    }

    public byte[] Seal(byte[] payload)
    {
        var nonce = RandomNumberGenerator.GetBytes(12);
        var cipher = new BouncyCastleChaCha20Poly1305();
        cipher.Init(true, new AeadParameters(new KeyParameter(key), 128, nonce));
        var encrypted = new byte[cipher.GetOutputSize(payload.Length)];
        var len = cipher.ProcessBytes(payload, 0, payload.Length, encrypted, 0);
        cipher.DoFinal(encrypted, len);
        return nonce.Concat(encrypted).ToArray();
    }

    public byte[] Open(byte[] sealedPayload)
    {
        if (sealedPayload.Length < 12)
        {
            throw new InvalidOperationException("Encrypted payload is too small.");
        }

        var nonce = sealedPayload[..12];
        var encrypted = sealedPayload[12..];
        var cipher = new BouncyCastleChaCha20Poly1305();
        cipher.Init(false, new AeadParameters(new KeyParameter(key), 128, nonce));
        var output = new byte[cipher.GetOutputSize(encrypted.Length)];
        var len = cipher.ProcessBytes(encrypted, 0, encrypted.Length, output, 0);
        var finalLen = cipher.DoFinal(output, len);
        return output[..(len + finalLen)];
    }

    public static LocalDeviceIdentity MakeIdentity(string displayName)
    {
        var privateKey = new X25519PrivateKeyParameters(new SecureRandom());
        var privateBytes = new byte[32];
        privateKey.Encode(privateBytes, 0);
        var publicBytes = new byte[32];
        privateKey.GeneratePublicKey().Encode(publicBytes, 0);
        var identity = new DeviceIdentity(
            Guid.NewGuid().ToString().ToUpperInvariant(),
            displayName,
            DevicePlatform.windows,
            Convert.ToBase64String(publicBytes));
        return new LocalDeviceIdentity(identity, privateBytes);
    }
}

public static class LocalLinkHashes
{
    public static string Sha256Hex(byte[] data) =>
        Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();

    public static string PairingCode(string localPublicKey, string remotePublicKey)
    {
        var joined = string.Join("|", new[] { localPublicKey, remotePublicKey }.Order(StringComparer.Ordinal));
        var digest = SHA256.HashData(System.Text.Encoding.UTF8.GetBytes(joined));
        long value = 0;
        for (var i = 0; i < 4; i++)
        {
            value = (value << 8) | digest[i];
        }

        return (value % 1_000_000).ToString("D6");
    }
}
